### ISSUE TITLE
Port End-to-End tests to ginkgo/gomega

### ISSUE THREAD

**Issue by rrati** [kubernetes/kubernetes#3837]
Currently the end-to-end (e2e) tests run using a home grown test framework which may not provide all the functionality of an established testing framework and may require additional maintenance of the framework itself.  Ginkgo/Gomega seem to be a pretty complete Go testing frameworkl.

Some of the functionality ginkgo/gomega provides:
- test suite randomization
- parallel test execution
- custom reporting
- spec structuring similar to other testing frameworks

There's a lot more functionality provided which may or may not be useful.  Information about ginkgo/gomega can be found here: http://onsi.github.io/ginkgo/

Moving the e2e suite from the current framework to ginkgo/gomega would require porting the existing go tests and implementing any existing shell tests in go.  The porting of a test written in go from the existing framework to ginkgo/gomega is pretty straight forward.  Unless some helper functions are required it is mostly a copy of the test code into a ginkgo/gomega It stanza.  Logging should be changed from the current glog to either ginkgo/gomega By statements or standard fmt.Print*.  Logs related to failure should use the Fail method.

Some POC code that ports of a few of the tests to ginkgo/gomega can be found here:
https://github.com/rrati/kubernetes/tree/ginkgo/test/ginkgo

To run the tests:
1. cd test/ginkgo
2. GOPATH=$KUBE_ROOT/_output/local/go:$KUBE_ROOT/Godeps/_workspace:$GOPATH ginkgo
or 
GOPATH=$KUBE_ROOT/_output/local/go:$KUBE_ROOT/Godeps/_workspace:$GOPATH go test

Wheren KUBE_ROOT is the root of the kubernetes source tree

**Comment by timothysc** [comment]
+1 per discussions.

**Comment by filbranden** [comment]
I'm particularly interested in custom reporting to generate JUnit XML for Jenkins...

I have a commit importing ginkgo/gomega into the Godeps tree, I was planning to tweak cmd/e2e to start using it just for the reports (right now) to see if that is easy to do...

Let's try to coordinate efforts.

Cheers,
Filipe

**Comment by roberthbailey** [comment]
@lavalamp (in case he is interested while on leave)

One thing that hasn't been mentioned is that this will add a third party dependency to our e2e tests. Right now we don't have any non-standard package dependencies (see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/e2e.go#L17) but it's unclear to me whether this is because the code was in hack/ or whether it was a philosophical choice to keep it simple to run the tests (without needing to munge the GOPATH as we do once we add dependencies on other libraries).

**Comment by alex-mohr** [comment]
@davidopp as part of test team kickoff?

**Comment by alex-mohr** [comment]
@satnam6502 who's been go-ifying some bash tests too

**Comment by zmerlynn** [comment]
This also gets a +1 from me, from on the GVC.

**Comment by satnam6502** [comment]
Is there some kind of test team kickoff?

**Comment by alex-mohr** [comment]
> Is there some kind of test team kickoff?
I heard someone at Google is going to be working full-time on testing in the near term.

**Comment by timothysc** [comment]
@filbranden @alex-mohr @roberthbailey @zmerlynn @satnam6502 @thockin 

Questions: 
Is there a content owner for the e2e tests, who we should coordinate with?
We would like to proceed and perhaps just go for full conversion, are there issues here? 
Is there some division of labor that folks would care to coordinate on?

**Comment by zmerlynn** [comment]
(If anyone wants to follow along, this conversation is happening live on
irc)

**Comment by thockin** [comment]
If there is an owner it is Zach or Satnam :)

**Comment by zmerlynn** [comment]
Sorry, forgot to summarize the IRC convo. I believe we decided to go ahead
and get #3855 in, which happened last night, and that rrati and others
would work on porting the tests one by one after that went in.

Does that sound like the correct short summary?

**Comment by satnam6502** [comment]
As for owners, given all the excellent work that Zach has taken the initiative on and driven and given his knowledge and expertise I think he is the right person to be the owner.

**Comment by rrati** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/issues/3941

**Comment by rrati** [comment]
Here's a proposal for moving the remaining tests into ginkgo test suites:

networking
- TestNetwork
- TestClusterDNS
- TestImportantURLs
- TestKubernetesROService

events
- TestKubeletSendsEvent

pods (add to existing)
- TestLivenessHttp
- TestLivenessExec

replication
- TestBasic
- TestPrivate

Thoughts?
@filbranden @satnam6502 @roberthbailey @zmerlynn

**Comment by filbranden** [comment]
Ok with me I guess... I'm not crazy about `replication`, maybe `rc`? Not sure if those tests are the best representation of replication controller tests, on the other hand we can start adding to them later on.

**Comment by rrati** [comment]
@filbranden I agree they aren't the best representation of replication controller tests.  They also didn't seem to fit in very well in the other suites I proposed. Is there a better place for them?

**Comment by filbranden** [comment]
One small concern (now that I started working again on some pod tests) is that it used to be possible to run a single test using e.g. `-t TestPodHasServiceEnvVars` but now all I have is `-t Pods` which runs all the 3 Pods tests. Pretty soon that will be 5 Pods tests. So it's a bit awkward that I start having to run more than a single test on each execution...

Maybe we can extend `-t` to work with the description of a single test, but not sure if it's obvious how to use that since that's actually a full description sentence, so not sure I like that more than having a very explicit name for each test.

I like Ginkgo in general, just that seems to be a bit of a step backwards from where we were... Suggestions?

**Comment by zmerlynn** [comment]
Watch "TestNetwork". Right now it's called out specifically by another
script, so it may need it's own suite?

**Comment by filbranden** [comment]
So turns out using `-t "environment variables"` actually works to run what used to be `TestPodHasServiceEnvVars` but that's quite ad-hoc and depends on not introducing other It()s which use similar wording...

**Comment by rrati** [comment]
@zmerlynn We should be able to test a single spec in a suite.  If we want to keep that linkage we can leave TestNetwork alone for now ot update the caller to use the necessary identifier to just run the network test.

**Comment by rrati** [comment]
@filbranden If we were using the ginkgo command line we might be able to only execute the tests desired with --focus.  It would still rely on the test description though.

We could also use a keyword at the start of the description that is unique.  That feels kinda hackish, takes us back to the function name format, and certainly wouldn't provide english-like output in the test report. :)  Might be better than relying on a random string though.

**Comment by davidopp** [comment]
Sorry, yeah, I had signed up to work on soak tests, though I'm not sure that's supposed to be my focus anymore. Anyway, for this particular issue I agree @zmerlynn is the right assignee.

**Comment by roberthbailey** [comment]
I'd group TestClusterDNS & TestKubernetesROService as "service" tests rather than network tests. TestImportantURLs is really a backwards compatibility test rather than a network test. 

Have you looked at the remaining shell tests to see where they would fit into this grouping?

**Comment by rrati** [comment]
@roberthbailey Would it make sense to put TestImportantURLs into the "service" test suite rather than the network test suite?  It seemed more like a service than network really.

TestNetwork is bit fuzzy too.  I put it in a network suite, but most of it's magic seems to be done in the image that is used.  It'd be great to inspect what that is doing and better document the test.

I haven't looked at the remaining shell tests yet.  My first pass is to re-org the existing e2e tests.  If new tests would be better off in a different suite than what is existing then we can always create a new suite.

**Comment by zmerlynn** [comment]
TestImportantUrls is honestly a basic REST test. I'd just keep it in network, or potentially make it a versioning test of some sort (that we have no other tests for yet).

Some of this seems kind of bikeshed-y right now. I'd honestly feel free to proceed and we'll figure out the rough edges as we go forward. Porting the shell tests will show more. (We've got a lot of wood to chop here, folks.)

**Comment by roberthbailey** [comment]
I agree that we should proceed and we can reorganize later. Breaking the monolithic shell tests up into smaller independent units written in Go (however those end up being grouped) is a strict improvement over what we have right now.

**Comment by onsi** [comment]
I'm super excited to hear y'all are giving Ginkgo/Gomega a try :)

Ginkgo was written with exactly these kinds of integration tests in mind.  You might want to look into using `Eventually` and, maybe, writing custom matchers.  You'd get a lot of what you're trying to do in

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/util.go

for free.

Over at cloud foundry we're using Ginkgo to run integration tests against a deployed cluster:

https://github.com/pivotal-cf-experimental/vizzini

has a lot of useful tips/tricks that might be worth looking into.

**Comment by onsi** [comment]
Another thing: by not using the `ginkgo` CLI you're going to make it hard to support running parallel tests down the line (and running tests in parallel is one of the best ways to speed up slow integration suites).

If there are feature gaps in `ginkgo` that you'd like to see fixed I'd be happy to add them.  In particular, [it looks like you want to be able to re-run a test suite up-to N times](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/driver.go#L78) that would be pretty easy to support as a first class thing (Ginkgo already has `--untilItFails` which is similar).

tl;dr -- am happy to help, feel free to open issues at github.com/onsi/ginkgo :)

**Comment by filbranden** [comment]
@onsi Thanks for your support!

Yes I have been planning to look into having the tests run by `ginkgo` directly, I think I'm close but I'll need to spend a little more time to really make it work...

One part I miss a bit is being able to give tests names (or more structured names), I know there's "focus" but that's a bit fuzzy and so far I haven't been able to understand what exactly it's matching against... In any case, not a huge deal. If I think of anything that could be improved in Ginkgo I'll certainly open an issue in the project!

Thanks!
Filipe

**Comment by onsi** [comment]
@filbranden 

`--focus` searches through the [`FullTestText`](https://github.com/onsi/ginkgo/blob/master/ginkgo_dsl.go#L134) of each `It`

I agree that it can be hard to give tests more structured names.  I tend to work around this by putting leading "tags" in relevent `Describe`/`Context` blocks.  Something like:

```
Describe("{FOO}{SLOW} A very slow test that checks that Foo works", func() {
    It("does something slow...", func() {
        ...
    })
})
```

Then you can `ginkgo --focus='{FOO}'` or `ginkgo --skip='{SLOW}'`. You can also combine `--focus` and `--skip`:

`ginkgo --focus='{FOO}' --skip='{SLOW}'

or use regular expressions:

`ginkgo --focus='{FOO}|{BAR}'`

Anway - it's the best I've come up with though I'm open to suggestions (at which point we'd probably move the conversation over to Ginkgo).

**Comment by filbranden** [comment]
Tags sound great.

Maybe make it more structured as a `[]string` or add a new "verb" (like `Describe`, `It` etc.) that you could use to associate tags to existing tests?

Ideally a unique name, though if we have tags we can always use `name=TestThisAndThat` which is unique enough...

Agree about moving to Ginkgo, I'll open an issue there right now.

Thanks!
Filipe

**Comment by ihmccreery** [comment]
Can we close this?  Is this work done?

**Comment by filbranden** [comment]
Yes, as far as I can tell, this is completed. @zmerlynn can confirm for sure.

**Comment by zmerlynn** [comment]
:+1:
